metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/storage/BlockStorage.scala

```diff
@@ -151,13 +160,24 @@ class BlockStorage[N, A <: Agreement: Block](
         case Some((blockHash, queue)) =>
           parentToChildrenColl.read(blockHash).flatMap {
             case None =>
-              loop(queue, acc)
+              // Since we're not inserting an empty child set,
+              // we can't tell here if the block exists or not.
+              loop(queue, blockHash :: acc)
```

Comment:
If we forego adding `blockHash` to result here, then the final result could be slightly more concise:

```scala
    loop(Queue(blockHash), Nil).flatMap { result =>
      contains(blockHash).map {
        case true  => blockHash :: result
        case false => result
      }
    }
```

Reply:
I'll check but I think that would be different: if I don't add the `blockHash` where `parentToChildren.read(blockHash)` is `None` then I'll also not get in the result any descendant which doesn't have further children on its own, so, none of the leaves. 

Also didn't want to do another `contains` check unless it looked suspicious.

Reply:
Unfortunately this fails many tests, or I maybe be misunderstanding what you wanted to do.

Reply:
Ah, true. There's something l don't like about that check, but since it's an optimisation for `result.size == 1`, well, optimisations tend to look like that.

Reply:
How about this?
```scala
   loop(Queue(blockHash), Nil).flatMap {
      case result @ List(`blockHash`) =>
        result.filterA(contains)
      case result =>
        KVStoreRead[N].pure(result)
    }
```

Reply:
Yes, much more pleasant to the eye :+1: